Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm

```diff
@@ -0,0 +1,196 @@
+static NSString *originsKey = @"origins";
+
+void WebExtensionContext::getAll(WebPageProxyIdentifier identifier, CompletionHandler<void(Vector<String>, Vector<String>)>&& completionHandler)
+{
+    Vector<String> permissions, origins;
+
+    for (auto& permission : currentPermissions())
+        permissions.append(permission);
+
+    for (auto& matchPattern : currentPermissionMatchPatterns())
+        origins.append(matchPattern->string());
+
+    completionHandler(permissions, origins);
+}
+
+void WebExtensionContext::contains(WebPageProxyIdentifier identifier, HashSet<String> permissions, HashSet<String> origins, CompletionHandler<void(bool)>&& completionHandler)
+{
+    if (m_backgroundWebView.get()._page->identifier() != identifier) {
+        completionHandler(false);
+        return;
+    }
+
+    MatchPatternSet patterns;
+    for (auto& origin : origins) {
+        auto pattern = WebExtensionMatchPattern::getOrCreate(static_cast<NSString *>(origin));
+        if (!pattern || !pattern->isValid()) {
+            completionHandler(false);
+            return;
+        }
+
+        patterns.add(*pattern);
+    }
+
+    if (!hasPermissions(permissions, patterns)) {
+        completionHandler(false);
+        return;
+    }
+
+    completionHandler(true);;
+}
+
+void WebExtensionContext::request(WebPageProxyIdentifier identifier, HashSet<String> permissions, HashSet<String> origins, CompletionHandler<void(bool, String)>&& completionHandler)
+{
+    if (m_backgroundWebView.get()._page->identifier() != identifier) {
+        completionHandler(false, ""_s);
+        return;
+    }
+
+    HashSet<Ref<WebExtensionMatchPattern>> matchPatterns;
+    NSString *errorMessage;
+
+    if (!validatePermissionsDetails(permissions, origins, matchPatterns, @"request()", &errorMessage)) {
+        completionHandler(false, errorMessage);
+        return;
+    }
+
+    if (!verifyRequestedPermissions(permissions, origins, matchPatterns, @"request()", &errorMessage)) {
+        completionHandler(false, errorMessage);
+        return;
+    }
+
+    if (hasPermissions(permissions, matchPatterns)) {
+        completionHandler(true, ""_s);
+        return;
+    }
+
+    // FIXME: <https://webkit.org/b/250135> Create and show access alert for requested permissions.
+}
+
+void WebExtensionContext::remove(WebPageProxyIdentifier identifier, HashSet<String> permissions, HashSet<String> origins, CompletionHandler<void(bool, String)>&& completionHandler)
+{
+    if (m_backgroundWebView.get()._page->identifier() != identifier) {
+        completionHandler(false, ""_s);
+        return;
+    }
+
+    HashSet<Ref<WebExtensionMatchPattern>> matchPatterns;
+    NSString *errorMessage;
+
+    if (!validatePermissionsDetails(permissions, origins, matchPatterns, @"remove()", &errorMessage)) {
+        completionHandler(false, errorMessage);
+        return;
+    }
+
+    if (!verifyRequestedPermissions(permissions, origins, matchPatterns, @"remove()", &errorMessage)) {
```

Comment:
Currently, in Safari we don't report back an error message and just fail silently to match Chrome's behavior. Although, I think passing back an error message in this case might be helpful to developers.

Reply:
Safari does report the error, it is doing it on the web process side via `validatePermissionDetails()`, not in the UI process.